### PR TITLE
DragPanel: Panel which can be reordered with drag and drop

### PR DESCRIPTION
**Pull request by vidartf**
This is a general purpose panel for using drag and drop to reorder widgets. The base implementation only allows for reordering widgets that are directly under the panel, but is implemented in such a way to be easily extensible in subclasses for more advanced behavior.

Example extension for allowing drag and drop reordering of cells (ref #641) would:
 - Override `move()` to change the order in the model instead of the order of the widgets.

OR:
 - Override `addMimeData()` to add `nbformat.ICell` to metadata bundle
 - Override `processDrop()` to move/copy the dropped cell.
 - Optionally set `allowDropsFromExternalSource` to true to allow copy/move of cells between panels.

The code is based on initial suggestion by @blink1073 in jupyter-attic/jupyter-js-notebook#189

### CODE REVIEW

**Comment by vidartf** [comment]
Note that the styling / UX for the end user is basically the same as in jupyter-attic/jupyter-js-notebook#189, except the easier way of specifying drag handles (elements that can be used to initiate a drag).

**Comment by blink1073** [comment]
Thanks for this, @vidartf!

**Comment by ellisonbg** [comment]
One thing we need to watch. A single widget, such as a cell, may need to have different drag targets all over the app. For example, a use may want to drag a cell:

* To a different location in a notebook
* To a different notebook
* To a text editor (and have the cell content get inserted into the text editor)

We need to make sure the design of this drag panel allows for those usage cases.

Awesome though!!!

**Comment by vidartf** [comment]
@ellisonbg I tried to have such scenarios in mind, which is where the serialized mime data is useful. The receiving panels would have to allow external drops, and then override `findDropTarget()` to scan the mime data of incoming drops for matching mime type (`null` return indicates no drop target could be found).

I tried to write docstrings that are helpful for subclassing, but any feedback on the docs is of course very welcome as well :)

**Comment by vidartf** [comment]
To try to make its functionality more clear, I attempted to replaced the drag and drop behavior in the file browser with the new code. In the attempt, I realized I would have to separate the drag and the drop functionality, so I refactored the components. Hopefully the code didn't become too hard to follow because of this.

At the end, I hope the code in the file browser became easier to maintain, and that this can serve as an example for how to use the drag/drop widgets.

**Comment by ellisonbg** [comment]
We talked a bit more about this and think that this adds a bit too much complexity when using the underlying API isn't much more complex. For now we would like to keep using the raw drag and drop API from phosphor and see more what common patterns emerge from that usage. Once we do that we could see if there are simpler ways of abstracting those common usage cases.

Given that, are you OK with us closing this?

**Comment by vidartf** [comment]
Sure, I'm just curious as to whether you felt the interface for subclassing was too complex, or if it was the underlying implementation (or both). It would of course be possible to simplify this code, but depending on what was cut, it might no longer work for my use cases (in which case I would have to keep a separate implementation anyway).

**Comment by blink1073** [comment]
Hi @vidartf, thank you for making the effort here.  For me, it exposed the fact that that the handling of the phosphor drag-drop API is very context-specific and the amount of boiler plate required is not overly cumbersome compared to the mental overhead of the myriad of different layout/panel combinations that came out of this, along with a necessarily diminished flexibility.

**Comment by vidartf** [comment]
@blink1073, yeah that makes sense. Drag/drop helpers like this are best implemented as mixins, which Typescript does not currently support well. If Typescript get support for it in the future, it might be worthy of a second attempt (maybe with a simpler subclassing interface).
